typing/typecore.ml

```diff
@@ -8001,15 +8026,18 @@ and type_format loc str env =
   with Failure msg ->
     raise (Error (loc, env, Invalid_format msg))
 
-and type_option_some env expected_mode sarg ty ty0 =
-  let ty' = extract_option_type env ty in
-  let ty0' = extract_option_type env ty0 in
+and type_option_some env mpath expected_mode sarg ty ty0 =
+  let ty' = extract_option_type env mpath ty in
+  let ty0' = extract_option_type env mpath ty0 in
   let alloc_mode, argument_mode = register_allocation expected_mode in
   let arg = type_argument ~overwrite:No_overwrite env argument_mode sarg ty' ty0' in
-  let lid = Longident.Lident "Some" in
-  let csome = Env.find_ident_constructor Predef.ident_some env in
+  let lid, csome = (match mpath with
+    | Stdlib_option -> (Longident.Lident "Some",
+      Env.find_ident_constructor Predef.ident_some env)
+    | Stdlib_or_null -> (Longident.Lident "This",
+      Env.find_ident_constructor Predef.ident_this env)) in
```

Comment:
Nit: `in` in multi-line let expressions should be on its own line, with the same indentation as `let`.